extending filesystem support for Hermit

stlankes opened the pull request:
Extending `std` to create, change and read a directory for Hermit.

Hermit is a tier 3 platform and this PR changes only files, wich are related to the tier 3 platform.

simonschoening committed:
Extending filesystem support for hermit-os

stlankes commented:
Cc: @mkroening

thomcc commented on library/std/src/sys/hermit/fs.rs:
```diff
@@ -119,75 +165,127 @@ impl FileTimes {
+                        } else {
+                            #[allow(deref_nullptr)]
+                            {
+                                ptr::addr_of!((*ptr::null::<dirent>()).$field)
```
This is very gnarly, even if the branch is never taken....

thomcc reviewed (commented):
I'll give it a more thorough pass later this weekend. It mostly looks fine.

thomcc commented on library/std/src/sys/hermit/fs.rs:
```diff
@@ -119,75 +165,127 @@ impl FileTimes {
+                    }
+                };
+
+                macro_rules! offset_ptr {
```
I'm not sure why this macro exists. Most/all of the use sites could be fine with another construct (either addr_of, or just `(*p).field`), no?

The only place I see something like it being needed is for the name field of the dirent (since it might not be entirely there) but at that point you can just use offset_of and casting, no?

stlankes commented:
Hm, you are right. I will revise it.

mkroening commented:
The `offset_of` macro seems to have its origin and explenation here: [sys/unix/fs.rs#L705-L736](https://github.com/rust-lang/rust/blob/26eeea6a1b94418a3f513c0cd81f00c4c2966dfd/library/std/src/sys/unix/fs.rs#L705-L736).

Since we control both sides of this API, we might be able to avoid having to use it.

thomcc commented:
I see, so that code is mostly copied from sys::unix? Hm... That's probably fine, then. It would be nice if it could be somehow reused, admittedly.

stlankes commented:
Hm, I agree, that this will be useful. However, the handling of error numbers are different and als a few system calls (e.g. `readdir`) have a different. interface. Maybe I overlook something. But I think that this isn't so easy.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6311415787/job/17135355755) [(plain)](https://github.com/rust-lang/rust/commit/d577cb6d149ca62421f6fb9243c8a62e408a5a5b/checks/17135355755/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=stlankes
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_a599f9bc-fec7-4955-af5a-c6454e1e61b5
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=683d4667171cb31f81d5b84b484b1a8d493c55ea
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_a599f9bc-fec7-4955-af5a-c6454e1e61b5
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_a599f9bc-fec7-4955-af5a-c6454e1e61b5
GITHUB_TRIGGERING_ACTOR=stlankes
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/115984/merge
GITHUB_WORKFLOW_SHA=683d4667171cb31f81d5b84b484b1a8d493c55ea
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_18_X64=/opt/hostedtoolcache/go/1.18.10/x64
---
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/d23062b5be6bbf5be15226c7f497c89cf7d3d6e1/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-d23062b5be6bbf5be15226c7f497c89cf7d3d6e1-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
##[group]Building stage0 tool expand-yaml-anchors (x86_64-unknown-linux-gnu)
    Updating crates.io index
error: the lock file /checkout/Cargo.lock needs to be updated but --locked was passed to prevent this
If you want to try to generate the lock file without accessing the network, remove the --locked flag and use --offline instead.
  local time: Tue Sep 26 10:35:43 UTC 2023
  network time: Tue, 26 Sep 2023 10:35:43 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

stlankes commented:
I rebased my branch. I would assume that my changes have no influence on the CI. Does the CI have some problems?

bors commented:
:umbrella: The latest upstream changes (presumably #116956) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

mkroening commented:
Rebased on latest `master`.

@thomcc, are you okay with the macro for now?

bors commented:
:umbrella: The latest upstream changes (presumably #118046) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #117285) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

stlankes committed:
revise interface to read directory entries

The new interface has some similarities to Linux system call
getdents64. The system call reads several dirent64 structures.
At the end of each dirent64 is stored the name of the file.
The length of file name is implictly part of dirent64 because
d_reclen contains size of dirent64 plus the length of the file
name.

stlankes commented:
@thomcc We revised the system call to read directory entries. I hope that interface is now more understandable.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7566797163/job/20604752046) [(plain)](https://github.com/rust-lang/rust/commit/43b64f78eccf33611b0c995e0f0532563cf6574a/checks/20604752046/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=stlankes
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_005940eb-d180-4f1c-9bcc-501685acfc5d
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=3f5d8e3ebe88e2435ab15e1d8808d044f5fd8754
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_005940eb-d180-4f1c-9bcc-501685acfc5d
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_005940eb-d180-4f1c-9bcc-501685acfc5d
GITHUB_TRIGGERING_ACTOR=stlankes
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/115984/merge
GITHUB_WORKFLOW_SHA=3f5d8e3ebe88e2435ab15e1d8808d044f5fd8754
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
Removing intermediate container b6aa4af8679f
 ---> 86e9ffbefeb2
Step 6/10 : COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
 ---> 47bff952235b
Step 7/10 : RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
Collecting binaryornot==0.4.4
  Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
Collecting boolean-py==4.0
  Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180123 sha256=f323ccf11d14c5b11f6d5e70edb46e391f4b49a5df5c7d922224477ad8ee15c5
  Stored in directory: /tmp/pip-ephem-wheel-cache-t30wgac9/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
    Can't uninstall 'setuptools'. No files were found to uninstall.
Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
Collecting virtualenv
  Downloading virtualenv-20.25.0-py3-none-any.whl (3.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 26.5 MB/s eta 0:00:00
Collecting distlib<1,>=0.3.7
  Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
Collecting filelock<4,>=3.12.2
  Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
  Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
Collecting platformdirs<5,>=3.9.1
  Downloading platformdirs-4.1.0-py3-none-any.whl (17 kB)
Installing collected packages: distlib, platformdirs, filelock, virtualenv
Successfully installed distlib-0.3.8 filelock-3.13.1 platformdirs-4.1.0 virtualenv-20.25.0
Removing intermediate container 140ed29405bc
 ---> 947647ec6308
Step 8/10 : COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
 ---> 52885dbfaea0
 ---> 52885dbfaea0
Step 9/10 : COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/
 ---> 849ef9b3f620
Step 10/10 : ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
Removing intermediate container 6e134b656b0b
 ---> 9e93a7a80dbb
Successfully built 9e93a7a80dbb
Successfully tagged rust-ci:latest
Successfully tagged rust-ci:latest
##[endgroup]
Built container sha256:9e93a7a80dbbab146be504100af2120290621f4bc3b069cccfb61de30822681d
Uploading finished image sha256:9e93a7a80dbbab146be504100af2120290621f4bc3b069cccfb61de30822681d to https://ci-caches.rust-lang.org/docker/139141a603ff222e98d20ef66f0e4eaedfc4b373e41d302ecff577f90ee0c6cf0cab092d18a8f6a073c88b362b9a680d3b2b6e5c828faaf25d1b853eabb78b1a
IMAGE          CREATED              CREATED BY                                      SIZE      COMMENT
9e93a7a80dbb   1 second ago         /bin/sh -c #(nop)  ENV SCRIPT=TIDY_PRINT_DIF…   0B        
52885dbfaea0   2 seconds ago        /bin/sh -c #(nop) COPY file:078ea1d11e7b7cda…   367B      
947647ec6308   3 seconds ago        |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.9MB    
47bff952235b   10 seconds ago       /bin/sh -c #(nop) COPY file:ac591dd6bc5afa66…   5.33kB    
86e9ffbefeb2   12 seconds ago       |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.1MB    
---
<missing>      6 days ago           /bin/sh -c #(nop)  LABEL org.opencontainers.…   0B        
<missing>      6 days ago           /bin/sh -c #(nop)  ARG LAUNCHPAD_BUILD_ARCH     0B        
<missing>      6 days ago           /bin/sh -c #(nop)  ARG RELEASE                  0B        

<botocore.awsrequest.AWSRequest object at 0x7f1cef18f3d0>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
https://ci-caches.rust-lang.org/docker/139141a603ff222e98d20ef66f0e4eaedfc4b373e41d302ecff577f90ee0c6cf0cab092d18a8f6a073c88b362b9a680d3b2b6e5c828faaf25d1b853eabb78b1a
sha256:9e93a7a80dbbab146be504100af2120290621f4bc3b069cccfb61de30822681d
---
DirectMap4k:      186304 kB
DirectMap2M:     5056512 kB
DirectMap1G:    13631488 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished dev [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/2457c028c907b43511445677fe39b2dbf679a0cf/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-2457c028c907b43511445677fe39b2dbf679a0cf-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished release [optimized] target(s) in 25.04s
##[endgroup]
fmt check
##[error]Diff in /checkout/library/std/src/sys/pal/hermit/fs.rs at line 522:
         if readlen > 0 {
             // shrink down to the minimal size
             vec.resize(readlen.try_into().unwrap(), 0);
-            unsafe { abi::close(fd_raw); }
+            unsafe {
+                abi::close(fd_raw);
             break;
         }
 
 
##[error]Diff in /checkout/library/std/src/sys/pal/hermit/fs.rs at line 529:
         // if the buffer is too small, getdents64 returns EINVAL
         // otherwise, getdents64 returns an error number
         if readlen != (-abi::errno::EINVAL).into() {
-            unsafe { abi::close(fd_raw); }
+            unsafe {
+                abi::close(fd_raw);
             return Err(Error::from_raw_os_error(readlen.try_into().unwrap()));
         }
 
 
##[error]Diff in /checkout/library/std/src/sys/pal/hermit/fs.rs at line 539:
         // 1 MB for directory entries should be enough
         // stop here to avoid an endless loop
         if sz > 0x100000 {
-            unsafe { abi::close(fd_raw); }
+            unsafe {
+                abi::close(fd_raw);
+            }
             return Err(Error::from(ErrorKind::Uncategorized));
     }
     }
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/library/std/src/sys/pal/hermit/args.rs" "/checkout/library/std/src/sys/pal/hermit/thread_local_dtor.rs" "/checkout/library/std/src/sys/pal/hermit/memchr.rs" "/checkout/library/std/src/sys/pal/hermit/futex.rs" "/checkout/library/std/src/sys/pal/hermit/mod.rs" "/checkout/library/std/src/sys/pal/hermit/fs.rs" "/checkout/library/std/src/sys/pal/hermit/stdio.rs" "/checkout/library/std/src/sys/pal/hermit/fd.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Thu Jan 18 07:44:20 UTC 2024
  network time: Thu, 18 Jan 2024 07:44:20 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

thomcc commented:
I'm going to be away for a few months, so I'm rerolling my PRs so that folks don't have to wait for me. Sorry/thanks.

r? libs

bors commented:
:umbrella: The latest upstream changes (presumably #121800) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

workingjubilee commented on library/std/src/sys/pal/hermit/fd.rs:
```diff
@@ -48,6 +48,11 @@ impl FileDesc {
     }
+
+    pub fn fstat(&self, stat: *mut abi::stat) -> io::Result<()> {
+        cvt(unsafe { abi::fstat(self.fd.as_raw_fd(), stat) })?;
```
...Why did I have to trace through three different directories just to discover that this is referring to the `hermit_abi` crate? That is so confusing.

stlankes replied:
Excuse me for my late response. I am a little bit confused. Why do you have trace three different directories? Do you mean I should move `fstat` from from `fd.rs` to `fs.rs`?

stlankes replied:
Or do you want that rename `abi` to `hermit-abi`? Maybe it is easier to understand...

workingjubilee replied:
Ah, I was a bit confused by the levels of indirection in general, but yes, I think probably using `hermit_abi` might be a bit clearer.

stlankes replied:
Ok, I would prefer to create an additional PR for these changes because I have to touch every file in `library/std/src/sys/pal/hermit/`. In addition, it doesn't belong to this PR.

bors commented:
:pushpin: Commit b1c7804c684301213775a4f490605302aeec7789 has been approved by `m-ou-se`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=m-ou-se b1c7804c684301213775a4f490605302aeec7789 -->
<!-- homu: {"type":"Approved","sha":"b1c7804c684301213775a4f490605302aeec7789","approver":"m-ou-se","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
